Thread Safety Analysis: Support warning on taking address of guarded variables

melver opened the pull request:
**Note:** This PR has been superseded by https://github.com/llvm/llvm-project/pull/127396, which takes a different approach, but coverage should be comparable with less risk of false positives.

----------

Add the optional ability, via `-Wthread-safety-addressof`, to warn when taking the address of guarded variables.

The motivation is to avoid false negatives in large C codebases, where data structures are typically implemented through helpers that take pointers to instances of a data structure. While taking the address itself does *not yet constitute a race* (in the presence of missing locking), placing the requirement on the pointer-recipient to obtain locks to access the pointed-to data is most likely poor style.

When passing pointers to functions, the additional warning is analogous to warning about passing reference to functions. However, unlike the existing analysis for references, the analysis is unable to track pointers precisely. Given the trade-offs, the user needs to explicitly decide if the remaining false positives and negatives are an acceptable cost for detecting potentially unsafe passing of pointers to guarded variables. Given that existing codebases using `-Wthread-safety` likely have cases where taking the pointer to a guarded variable is benign, the feature is *not* enabled by default but requires explicit opt-in.

melver committed:
Thread Safety Analysis: Support warning on obtaining address of guarded variables

Add the optional ability, via `-Wthread-safety-addressof`, to warn when
obtaining the address of guarded variables.

This is required to avoid false negatives in large C codebases, where
data structures are typically implemented through helpers that take
pointers to instances of a data structure.

We also argue that, while obtaining the address itself does not yet
constitute a potential race (in the presence of missing locking),
placing the requirement on the pointer-recipient to obtain locks to
access the pointed-to data is most likely poor style. This is analogous
to passing C++ references to guarded variables, which produces warnings
by default.

Given that existing codebases using `-Wthread-safety` likely have cases
where obtaining the pointer to a guarded variable is benign, the feature
is not enabled by default but requires explicit opt-in.

melver commented:
For additional background: I'm trying to work out how to bring -Wthread-safety to the Linux kernel. Since -fexperimental-late-parse-attributes made the feature practical for complex C structs, I started to look at a strategy to enable the feature. 

The current strategy is to create an opt-in mechanism (per subsystem), and since the constraints imposed by Wthread-safety already require subtly restructuring most code that would opt-in, I think having Wthread-safety-addressof part of the initial constraints imposed by Wthread-safety would be very helpful.

In converting one subsystem I maintain, I was disappointed that things like `list_add(&note, &guarded_list)` aren't caught. In general, I believe that the situation is analogous to C++ reference passing, and thus -Wthread-safety-addressof would be the right choice for new C projects.

aaronpuchert commented:
We had a discussion on https://reviews.llvm.org/D52395 that might be relevant here. To quote @delesley:

>  When you pass an object to a function, either by pointer or by reference, no actual load from memory has yet occurred. Thus, there is a real risk of false positives; what you are saying is that the function *might* read or write from protected memory, not that it actually will.

Another aspect is that we don't check if the lock is still held when we dereference the pointer, so there are also false negatives:
```c++
Mutex mu;
int x GUARDED_BY(mu);

void f() {
  mu.Lock();
  int *p = &x;
  mu.Unlock();
  *p = 0;
}
```
You use the analogy of C++ references. Interestingly, we don't warn on "taking a reference", partly because no such thing exists. (You can only initialize something of reference type with an expression.) We warn on passing a variable by reference *to a function*, or in other words, initializing a parameter of reference type with a guarded variable. (Since https://reviews.llvm.org/D153131, we also warn on returning a reference when the lock is no longer held after return. Note that the initialization of the reference might still happen under lock!)

However, we also track references in a way that pointers are not tracked. The reason is probably that references are immutable (that is, `T&` is semantically the same thing as `T* const`).
```c++
void g() {
  mu.Lock();
  int &ref = x;
  mu.Unlock();
  ref = 0;  // warning: writing variable 'x' requires holding mutex 'mu' exclusively
}
```
If you "take the reference" outside of the lock and do the assignment inside, there is no warning. Because the assignment is what needs the lock, not taking the address.

However, with functions it's somewhat reasonable to assume that the function accesses through the pointer, and that the pointer doesn't escape. (This is of course not always true, but often enough.) So I wonder whether we should restrict this to pointers passed as function arguments.

But if the number of false positives is manageable, we can also leave it like you implemented it. Did you try it out on some code base that uses thread safety annotations?

melver commented:
> We had a discussion on https://reviews.llvm.org/D52395 that might be relevant here. To quote @delesley:
> 
> > When you pass an object to a function, either by pointer or by reference, no actual load from memory has yet occurred. Thus, there is a real risk of false positives; what you are saying is that the function _might_ read or write from protected memory, not that it actually will.

Right. Though in the absence of better pointer tracking / alias analysis, I believe there's no way to avoid resulting false positives with pointers. It's something that a user of Wthread-safety-addressof would opt-in explicitly - documentation needs elaboration on this perhaps.

> Another aspect is that we don't check if the lock is still held when we dereference the pointer, so there are also false negatives:
> 
> ```c++
> Mutex mu;
> int x GUARDED_BY(mu);
> 
> void f() {
>   mu.Lock();
>   int *p = &x;
>   mu.Unlock();
>   *p = 0;
> }
> ```

Good point - though I'd prefer few false negatives + false positives over no checks at all. It's a tradeoff, and therefore we definitely shouldn't enable Wthread-safety-addressof by default.

> You use the analogy of C++ references. Interestingly, we don't warn on "taking a reference", partly because no such thing exists. (You can only initialize something of reference type with an expression.) We warn on passing a variable by reference _to a function_, or in other words, initializing a parameter of reference type with a guarded variable. (Since https://reviews.llvm.org/D153131, we also warn on returning a reference when the lock is no longer held after return. Note that the initialization of the reference might still happen under lock!)
> 
> However, we also track references in a way that pointers are not tracked. The reason is probably that references are immutable (that is, `T&` is semantically the same thing as `T* const`).
> 
> ```c++
> void g() {
>   mu.Lock();
>   int &ref = x;
>   mu.Unlock();
>   ref = 0;  // warning: writing variable 'x' requires holding mutex 'mu' exclusively
> }
> ```
> 
> If you "take the reference" outside of the lock and do the assignment inside, there is no warning. Because the assignment is what needs the lock, not taking the address.

Good points, and I missed that reference handling is much better in this regard.

> However, with functions it's somewhat reasonable to assume that the function accesses through the pointer, and that the pointer doesn't escape. (This is of course not always true, but often enough.) So I wonder whether we should restrict this to pointers passed as function arguments.

That's a reasonable option to make it more conservative, although I'm not sure it's necessary (yet).

> But if the number of false positives is manageable, we can also leave it like you implemented it. Did you try it out on some code base that uses thread safety annotations?

I'm working on bringing Wthread-safety to the Linux kernel. The strategy chosen is to convert one subsystem (or single TU) at a time, based on an opt-in basis. Since Linux already has a rather special dialect of C, those that would choose to opt in their TUs would opt into another variant of Linux's C dialect (one with Wthread-safety enabled).

My experience in converting a subsystem I maintain is that applying Wthread-safety already requires small refactors to avoid false positives, and I'm not opposed to add additional constraints via Wthread-safety-addressof. Without additional checking of "obtaining address of guarded variables", I find that Wthread-safety barely provides benefits, as passing pointers to datastructures is so common.

Resolving this will improve the usefulness, and ultimately the chances of us succeeding to upstream this to the Linux kernel.

I don't think we will ever convert 100% of the kernel to use Wthread-safety, but the plan is that willing subsystem maintainers opt in and deal with warnings produced by Wthread-safety. Having Wthread-safety-addressof will add additional coverage. For cases where it produces too many false positives, I intend to add a simple option that can enable/disable addressof checking for individual TUs. Or if almost all places where obtaining the address of a particular lock-guarded variable is _not_ actually under a lock (but still correct), leaving off `guarded_by` might simply be the right choice.

melver committed:
Update documentation wording on false positives and negatives

Update documentation wording to be clearer about false positives and
negatives of Wthread-safety-addressof.

melver committed:
Change terminology "obtaining address of" -> "taking address of"

The wording "take address of" or "taking address of" is more common
across the codebase:

% git grep 'tak\(e\|ing\) address of' | wc
    107    1513   16080

Whereas the wording "obtaining address of" has no occurrence. Use the
more common wording.

melver commented:
FWIW, the Linux kernel integration (draft, WIP) currently lives here: https://github.com/google/kernel-sanitizers/tree/cap-analysis
It currently enables -Wthread-safety-addressof if available. Thus far, I have not found any false positives due to -Wthread-safety-addressof in the 2 subsystems I converted over (more to follow).

And I want to re-iterate that without -Wthread-safety-addressof, the feature's coverage is significantly reduced, and I predict it to be one of the first complaints in later review.

Kindly take another look.

cc @bvanassche

nickdesaulniers commented:
Looks like the kernel patches [use](https://github.com/google/kernel-sanitizers/commit/81883e1fa377d866c288192b6eb8334bcf10f38f) `-fexperimental-late-parse-attributes`? :(

melver commented:
> Looks like the kernel patches [use](https://github.com/google/kernel-sanitizers/commit/81883e1fa377d866c288192b6eb8334bcf10f38f) `-fexperimental-late-parse-attributes`? :(

Yeah, no way around it if we want to have guarded_by in structs work properly. If the problem is cost, still planning to allow it to be disabled via Kconfig option.

nickdesaulniers commented:
Can any of the members in the structs be reorganized to put the mutex member declaration BEFORE the members they guard? Probably not always, but perhaps that's possible for most structures?

melver commented:
> Can any of the members in the structs be reorganized to put the mutex member declaration BEFORE the members they guard? Probably not always, but perhaps that's possible for most structures?

It's an option I considered, but I can already hear "what is this crap ... NACK". In many cases it might be possible, but where data layout or cacheline organization is important for performance, definitely not an option.

nickdesaulniers commented:
Doesn't `counted_by` have the same requirements? If not, why does guarded_by?

melver commented:
> Doesn't `counted_by` have the same requirements? If not, why does guarded_by?

It does, AFAIK.

> Sure, I can imagine where there might be pushback, but perhaps its worthwhile to see how far you can get with either marking structs that don't need `-fexperimental-late-parse-attributes`, or propose such member declaration re-orderings. At least that would let you land something upstream and begin protecting the kernel/finding bugs.

Fair point, in the worst case individual TUs could still add the flag if needed.

nickdesaulniers commented:
> > Doesn't `counted_by` have the same requirements? If not, why does guarded_by?
> 
> It does, AFAIK.

But counted_by _isn't_ using `-fexperimental-late-parse-attributes` IIUC. Do they have the same problem then of needing to reorder member declarations? cc @kees @isanbard @bwendling

bwendling commented:
Our current `counted_by` implementation is for flexible array members only, which are always at the end of the struct (or supposed to be). When we start applying them to pointers in structs (coming soon), we'll need to use that flag. However, I think that Apple is __finally__ starting to release some of their features, and I hope that the flag will go away as it'll always be on.

melver committed:
fixup! Change terminology "obtaining address of" -> "taking address of"

ramosian-glider reviewed (approved):
LGTM (we've discussed minor nits in the comments offline, Marco fixed them)

melver commented:
RFC posted to LKML: https://lore.kernel.org/lkml/20250206181711.1902989-1-elver@google.com

Depending on how things go, I intend to commit this within a few weeks unless there are objections, so that folks that want to already use this, can do so by just grabbing latest Clang.

Thanks!

aaronpuchert commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -1983,11 +1983,21 @@ class ThreadSafetyReporter : public clang::threadSafety::ThreadSafetyHandler {
+      DiagID = diag::warn_addressof_requires_any_lock;
+      break;
+    default:
+      assert(false && "Only works for variables");
+      break;
```
Wouldn't `llvm_unreachable` be the standard idiom here?

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -1,7 +1,9 @@
 // RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 -Wthread-safety -Wthread-safety-beta -Wno-thread-safety-negative -fcxx-exceptions -DUSE_CAPABILITY=0 %s
 // RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 -Wthread-safety -Wthread-safety-beta -Wno-thread-safety-negative -fcxx-exceptions -DUSE_CAPABILITY=1 %s
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 -Wthread-safety -Wthread-safety-beta -Wno-thread-safety-negative -Wthread-safety-addressof -fcxx-exceptions -DUSE_CAPABILITY=1 -DCHECK_ADDRESSOF %s
```
Some goes here: the test is really big and I don't want a combinatorial explosion. We're also having `-beta` on by default, so there is no reason to have `-addressof` separately. (We have `-negative` separately because it's a different kind of analysis.)

aaronpuchert reviewed (commented):
I've taken a brief look at the [Linux kernel changes](https://github.com/google/kernel-sanitizers/compare/2014c95afecee3e76ca4a56956a936e23283f05b...cap-analysis). I didn't check how many variables are affected by `__rcu_guarded`, but otherwise there seem to be maybe two dozen guarded variables. If you really didn't encounter any false positives that's not bad.

However, I'd still prefer if we could mark this as experimental, because it's a bit coarse:
* The address-of operation isn't interesting by itself, it merely happens to be required to do pass-by-reference in C.
* The equivalent of passing a `pt_guarded_by` variable by value doesn't seem to warn.
* The actual access itself is not checked, whether explicit or assumed behind a function call.

In the long term, a better approach might be:
* Extend alias tracking to pointers, perhaps restricted to those that don't change value.
* Warn when passing pointers to guarded variables into other functions.

aaronpuchert commented:
> The equivalent of passing a `pt_guarded_by` variable by value doesn't seem to warn.

This inconsistency is probably my largest concern. If you have
```c
int x GUARDED_BY(mu);
int *p PT_GUARDED_BY(mu);
```
then `&x` should basically behave like `p`, and `x` like `*p`. But with the current implementation, that cannot be the case, because `p` is just a no-op.

That's why I think we should rather check function arguments, and warn on passing (1) the address of a guarded variable, (2) a pointee-guarded variable. There is already functionality that makes sure (1) and (2) work together: `checkAccess` and `checkPtAccess` call each other when necessary. This doesn't seem to handle unary `&` at the moment, but I suppose this would be easy to add. The check for passing arguments itself is in `examineArguments`. We currently only check reference type arguments:
```c++
    if (Qt->isReferenceType())
      checkAccess(*Arg, AK_Read, POK_PassByRef);
```
Here we might simply add a branch for pointer types with the appropriate `POK` that you already introduced (though we might need to rename it slightly).

This still doesn't cover local pointer variables, but here I guess we need a better understanding of the patterns and how much alias analysis we'd need to unravel them. Local pointer variables as obfuscation device are hopefully not common, so there is likely some more complex logic that would be intractable for us anyway. So I'd be curious to see some interesting examples.

melver committed:
fixup! Thread Safety Analysis: Support warning on obtaining address of guarded variables

melver commented on clang/docs/ThreadSafetyAnalysis.rst:
```diff
@@ -515,8 +515,18 @@ Warning flags
+  example, where the manipulation of common data structures is done via
+  functions that take pointers to instances of the data structure. Note,
+  however, that the analysis does not track pointers, and false positives *and*
+  negatives are still possible.
```
Made it depend on Wthread-safety-beta.

melver commented:
[...]
> Then we only need to make sure that `checkPtAccess` can look through `&`, as mentioned above. (Casts should already be unwrapped.) This might not even need a new flag, it's just closing a gap in the existing analysis.

Thanks for the suggestions!

I implemented both -- please see https://github.com/llvm/llvm-project/pull/127396.

aoates commented:
I'm very excited about this, as I have wanted it for many years for my C codebase, and TSA is not super useful in C without this!

One thought --- you could consider an attribute that could be put on pointer arguments to functions that says "yes, I dereference this and read or write it".  In a codebase that otherwise would have many false positives, you could annotate at least core data structures without having to turn it on for all address-of operations.

E.g
```void hashtable_insert(htbl_t* WRITES_POINTER table, ...)```

In the C codebase I desperately want this for, an annotation like that sprinkled in a couple key places would get you 80% of the benefit with much lower risk of false positives.

melver commented:
> I'm very excited about this, as I have wanted it for many years for my C codebase, and TSA is not super useful in C without this!

This PR is being superseded by https://github.com/llvm/llvm-project/pull/127396 (implementation changed completely) - we agreed to go with the more conservative approach, and at the same time also properly handle pt_guarded_by pointers. Wthread-safety-pointer is meant to be to pointer passing what Wthread-safety-reference is to C++ reference passing.

I'm leaving this PR up to keep the history. 

> One thought --- you could consider an attribute that could be put on pointer arguments to functions that says "yes, I dereference this and read or write it". In a codebase that otherwise would have many false positives, you could annotate at least core data structures without having to turn it on for all address-of operations.
> 
> E.g `void hashtable_insert(htbl_t* WRITES_POINTER table, ...)`
> 
> In the C codebase I desperately want this for, an annotation like that sprinkled in a couple key places would get you 80% of the benefit with much lower risk of false positives.

Perhaps if there are false positives this can be added. But initially I'd like to avoid it if at all possible. A similar precedent at least for passing references to functions exists with `Wthread-safety-reference` (for C++ references), so in a first iteration we have narrowed it down to effectively mirror Wthread-safety-reference but for pointers.

FWIW, something similar to explicit marking can be achieved with something like this (but yes, it's ugly):
```
void __hashtable_insert(htbl_t* table, ...);
#define hashtable_insert(table, ...) ({ (void)(*(table)); __hashtable_insert(table, ...); })
```
Although we might need this patch too, to handle things like `*&var`: https://github.com/llvm/llvm-project/pull/127396/commits/a70f021becb2888d6c2a63b2d1e619393a996058

aoates commented:
> > I'm very excited about this, as I have wanted it for many years for my C codebase, and TSA is not super useful in C without this!
> 
> This PR is being superseded by #127396 (implementation changed completely) - we agreed to go with the more conservative approach, and at the same time also properly handle pt_guarded_by pointers. Wthread-safety-pointer is meant to be to pointer passing what Wthread-safety-reference is to C++ reference passing.
> 
> I'm leaving this PR up to keep the history.

Ah great, I will follow along there. I might patch in the early version of this and give it a spin this week (I don't want to wait for it be merged :D)
> 
> > One thought --- you could consider an attribute that could be put on pointer arguments to functions that says "yes, I dereference this and read or write it". In a codebase that otherwise would have many false positives, you could annotate at least core data structures without having to turn it on for all address-of operations.
> > E.g `void hashtable_insert(htbl_t* WRITES_POINTER table, ...)`
> > In the C codebase I desperately want this for, an annotation like that sprinkled in a couple key places would get you 80% of the benefit with much lower risk of false positives.
> 
> Perhaps if there are false positives this can be added. But initially I'd like to avoid it if at all possible. A similar precedent at least for passing references to functions exists with `Wthread-safety-reference` (for C++ references), so in a first iteration we have narrowed it down to effectively mirror Wthread-safety-reference but for pointers.
> 
> FWIW, something similar to explicit marking can be achieved with something like this (but yes, it's ugly):
> 
> ```
> void __hashtable_insert(htbl_t* table, ...);
> #define hashtable_insert(table, ...) ({ (void)(*(table)); __hashtable_insert(table, ...); })

Yup, this is what I currently am playing with, but it's ugly (and requires double-evaluating the macro argument which screws up side effects).

> ```
> 
> Although we might need this patch too, to handle things like `*&var`: [a70f021](https://github.com/llvm/llvm-project/commit/a70f021becb2888d6c2a63b2d1e619393a996058)

aaronpuchert commented:
> One thought --- you could consider an attribute that could be put on pointer arguments to functions that says "yes, I dereference this and read or write it".

GCC [has such an attribute](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-access-function-attribute), independent of Thread Safety Analysis. But I don't think we have an equivalent.

However, as I wrote above, it's probably the rule and not the exception that pointers passed into a function are dereferenced at some point. So we're probably fine if we always warn. And as @melver pointed out, the experience with reference passing in C++ has been pretty good, and it's probably transferable to pointers.
